### ISSUE TITLE
[clang-15] wrong code at -O1 about -inline

### ISSUE THREAD

**Issue by Air111** [llvm/llvm-project#61713]
Hi,
I found clang-15 produce wrong code at -O1 on this test case:

```c
#include <stdint.h>
#include <stdio.h>
int8_t a = 0;
uint64_t c = 0;
int8_t d[3] = {0, 0, 0};
uint32_t o(uint32_t p, uint32_t q) { 
  return (q == 0)? (p) : (p % q); 
}
int main() {
  int8_t *g = &a;
  int8_t *i = &d[1];
  int32_t j = (g == i);
  printf("%d\n", j);
  int32_t l = 1;
  uint64_t *n = &c;
  *n = o(10, (uint8_t)(l | j));
  printf("%llu\n", c);
}
```

Use -fno-inline-functions can produce correct code.

```console
$ clang-15 -v
Ubuntu clang version 15.0.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/lib/llvm-15/bin
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.3.1
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.5.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/8
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Candidate multilib: x32;@mx32
Selected multilib: .;@m64
$ clang-15 -target i386-linux -O1 small.c -o small
$ ./small
0
1
$ clang-15 -target i386-linux -O1 -fno-inline-functions small.c -o small
$ ./small
0
0
```

**Comment by AMS21** [comment]
Interestingly clang-14, clang-16 and trunk generate the correct code [godbolt](https://godbolt.org/z/PKc3z1nzG)

**Comment by godmiaozi** [comment]
any update about this issue? i think we have encountered similar problems with clang15

**Comment by godmiaozi** [comment]
@nikic we are still using clang15 and it will take us a lot of time to switch llvm from 15 to16 for big projects, is it possible to identify what lead to this problem and make a patch for this on llvm 15?

**Comment by godmiaozi** [comment]
with git bisect, i figured out that this bug is fix by this commit d4940c0f3d433ca5bab66d1e6a5170059c38a10a
